### ISSUE TITLE
Tests for observation loss and resource leakage

### ISSUE THREAD

**Issue by waltermi** [mami-project/pathspider#183]
In the ECN plugin, the "ecn.negotiation" information of some measurements is missing when a lot of workers are running. Running fewer workers reduces the number of measurements with missing conditions. The problem also occured with some self-made chains.

**Comment by irl** [comment]
This makes me think that flows are being dropped within PATHspider, or packets are being dropped and never captured. Unless you can find a place in PATHspider that this is occuring where it shouldn't, this is not a bug. Just use a sensible number of workers.

**Comment by britram** [comment]
This framing may not have been the best way to ask this question, but it does seem that results are load-dependent and load-dependency in results is input size dependent, which suggests resource leakage. We don't have a good baseline or calibration for giving people who are trying to use PATHspider guidance on what "a sensible number of workers" is, and we should have that if we're going to close issues based on it.

So, I think what we need here is some idea of how many records go missing under which network, CPU, and memory conditions with what number for workers, which probably means profiling observer loss for a set of possible conditions on relatively constrained DO nodes.

**Comment by britram** [comment]
(I may have cycles for this in December, but not before.)

**Comment by mirjak** [comment]
Agree, we should probably add some more logging/tracking capabilities here as well...

**Comment by irl** [comment]
Logging/tracking is #155 probably. @mirjak if you have specific requests for logging, please add them there. Otherwise we can just go nuts adding logging (if we have a decent logging setup, logging should be near-zero cost).

**Comment by irl** [comment]
Let's say the criteria for this being closed is running benchmarks and documenting sensible worker counts for different setups? (Also making sure our defaults are not way off for most users)

**Comment by britram** [comment]
SGTM.

**Comment by britram** [comment]
n.b. if this is targeting 2.0 and assigned to me, it will cause 2.0 to drop late.

**Comment by mirjak** [comment]
I didn't check what's logged to far but know that there was a test but no observation would be really helpful here.

**Comment by irl** [comment]
@mirjak: Does #184 look good for that?

**Comment by mirjak** [comment]
Yes, however, not sure if not_observed should be a condition or just something to log... thanks!

**Comment by britram** [comment]
I'm pretty sure there's a not_observed condition, at least for ECN

**Comment by britram** [comment]
ah, i see I'm wrong.

in any case `not_observed` should *absolutely* be a condition -- metadata calibrating the measurement platform itself needs to be inline or it'll be ignored by later analysis.

**Comment by irl** [comment]
It's now included in the ECN plugin, which should make some automated benchmarking possible as we just need to analyse the output for unobserved flows.

**Comment by irl** [comment]
```
pspdr measure -w X -i eth0 ecn --connect tcp < /tmp/targetsp > /tmp/results
```

Rough numbers:

For 2320 jobs in a DO 2GB instance, 20 workers gives 20 losses. >30 workers causes all but around 300 to be lost.

It looks like DO has changed their networking as you now get a NAT IPv4 address which may be impacting the performance. This could also be Meltdown/Spectre microcode updates killing the CPU as I've seen from other cloud providers.

With the same task on my desktop (2x Intel E5640@2.67GHz, 24GB RAM, DSL connection), 2321 jobs with 100 workers only gives me 23 losses.

**Comment by irl** [comment]
This is a duplicate of #198
